src/core/thread/meshcop_dataset_manager.cpp

```diff
@@ -306,6 +314,196 @@ void DatasetManager::HandleGet(Coap::Header &aHeader, Message &aMessage, const I
     SendGetResponse(aHeader, aMessageInfo, tlvs, length);
 }
 
+ThreadError DatasetManager::SendSetRequest(otOperationalDataset &aDataset, const uint8_t *aTlvs, uint8_t aSize)
+{
+    ThreadError error = kThreadError_None;
+    Coap::Header header;
+    Message *message;
+    Ip6::Address leader;
+    Ip6::MessageInfo messageInfo;
+    Tlv tlv;
+    uint16_t data16;
+    uint32_t data32;
+    uint64_t data64;
+
+    mSocket.Open(&HandleUdpReceive, this);
+
+    for (size_t i = 0; i < sizeof(mCoapToken); i++)
+    {
+        mCoapToken[i] = static_cast<uint8_t>(otPlatRandomGet());
+    }
+
+    header.Init();
+    header.SetVersion(1);
+    header.SetType(Coap::Header::kTypeConfirmable);
+    header.SetCode(Coap::Header::kCodePost);
+    header.SetMessageId(++mCoapMessageId);
+    header.SetToken(mCoapToken, sizeof(mCoapToken));
+    header.AppendUriPathOptions(mUriSet);
+    header.AppendContentFormatOption(Coap::Header::kApplicationOctetStream);
+    header.Finalize();
+
+    VerifyOrExit((message = Ip6::Udp::NewMessage(0)) != NULL, error = kThreadError_NoBufs);
+    SuccessOrExit(error = message->Append(header.GetBytes(), header.GetLength()));
+
+    if (aDataset.mIsActiveTimestampSet)
+    {
+        tlv.SetType(OT_MESHCOP_TLV_ACTIVETIMESTAMP);
+        tlv.SetLength(sizeof(aDataset.mActiveTimestamp));
+        SuccessOrExit(error = message->Append(&tlv, sizeof(tlv)));
+        data64 = HostSwap64(aDataset.mActiveTimestamp);
+        SuccessOrExit(error = message->Append(&data64, sizeof(aDataset.mActiveTimestamp)));
```

Comment:
You should use the the specific tlv objects for each type (e.g. `ActiveTimestampTlv` in this case), rather than performing the TLV encoding yourself.